### PR TITLE
Add word-wrap support, with wrap length provided by the user

### PR DESCRIPTION
**Pull request by mostynb**
## What type of PR is this?

- feature

## What this PR does / why we need it:

The help text is difficult to read when the lines are long. This patch attempts to wrap long lines, if the user provides a terminal width to wrap at.

## Which issue(s) this PR fixes:

Updated version of #1044.

Refers to #287.

## Testing

Added a unit test (which can be expanded).

## Release Notes

```release-note
The help text is now wrapped to fit the terminal window.
```

### CODE REVIEW

File: internal/build/build.go
Comment (lynncyrin) [inline internal/build/build.go:197]: The binary size increasing and the new dependency (`golang.org/x/crypto`) concern me, can you rewrite this PR to remove those two changes?

Comment (mostynb) [reply internal/build/build.go:197]: I can experiment with some platform-specific code, and see what is possible. Would it be OK if this feature only works on linux/mac/windows (for a small subset of each that I can easily test)?

Comment (lynncyrin) [reply internal/build/build.go:197]: I needs to work on all of the platforms that we test on CI, and you don't need to support anything beyond that

File: termwidth_misc.go
Comment (mostynb) [inline termwidth_misc.go:5]: I need to double-check this.

File: internal/build/build.go
Comment (mostynb) [reply internal/build/build.go]: Even using the low-level `unix.IoctlGetWinsize`/`windows.GetConsoleScreenBufferInfo` calls, this increases the binary size by 0.1M. Though perhaps golang.org/x/sys/unix and golang.org/x/sys/windows are more likely to already be used by client programs than golang.org/x/crypto.

The only other option I can think of to not raise the binary size by this much would be for clients to figure out the terminal size themselves and provide it somehow (maybe set it as a new field in the App struct). What do you think?

Comment (lynncyrin) [reply internal/build/build.go]: There's 2 options here that stand out for me:

1. import the code for the `unix.IoctlGetWinsize` and `windows.GetConsoleScreenBufferInfo` functions so that we aren't importing the whole package
2. allow people to manually set the terminal size, like you mention

Either way, we definitely want to avoid increasing the binary size for this feature. At least, not in a pre-v3 world where can't make this feature opt-in. (see https://github.com/urfave/cli/issues/833#issuecomment-579869829 for context)

Comment (lynncyrin) [reply internal/build/build.go]: I would also say that it would probably be un-controversial to let people set the terminal size, and have a default terminal size of `80` as described here => https://github.com/urfave/cli/issues/287#issuecomment-320581746

Comment (mostynb) [reply internal/build/build.go]: I would rather not copy+paste all that subtle syscall code, so I think we should go for (2) - it also makes this feature testable.

**Comment by mostynb** [comment]
How does this look?

* Is it OK to add a new field to `helpPrinter`?
* The template probably needs tweaking to support wrapping more fields, but I don't know offhand which other fields to test.
* This makes no attempt to break long words.

File: go.mod
Comment (lynncyrin) [inline go.mod:9]: I hate to be a broken record, but please remove this change 🙏 I would rather this PR not add new requirements, even if they are only for testing.

Comment (mostynb) [reply go.mod]: No problem- done.

**Comment by lynncyrin** [comment]
I see you just force pushed, so I'm not sure what changed? Can you re-request my review when this is ready for eyes again, and please not force push until after the PR is ready to merge?

**Comment by mostynb** [comment]
> I see you just force pushed, so I'm not sure what changed? Can you re-request my review when this is ready for eyes again, and please not force push until after the PR is ready to merge?

Sorry about that- I squashed the commit in an attempt to make it easy to verify that no new modules were added.

I have re-pushed the branch with [the original commit](https://github.com/urfave/cli/pull/1119/commits/be03ef5e9b4f9a8300e2e3ed5dd972c7840cbec0) followed by a [fixup that removes the newly added module](https://github.com/urfave/cli/pull/1119/commits/7c8032c545db2da51ab2ec4d85f567e5b71009ce).

**Comment by lynncyrin** [comment]
I'll try to give this a look soon!

**Comment by ericmillin** [comment]
@lynncyrin Would love to see this merged.

**Comment by lynncyrin** [comment]
Oops 😄

File: template.go
Comment (lynncyrin) [inline template.go:10]: ❓ &nbsp; _(blocking question)_ what do the 6 and 3 here represent???

**Review by coilysiren — DISMISSED** [review]
I have 1 blocking question, since I'm not following the `template.go` changes @mostynb

File: template.go
Comment (mostynb) [reply template.go:10]: They are arguments to the functions `wap` and `offset`, in the funcMap applied to the template.

`wrap` takes a string and an offset to indent wrapped lines, and returns a wrapped string (the second line onwards are all indented by the offset. `3` is the indent level used in the template.

`offset` returns the length of a string plus a given number, and is used to calculate an offset for the description field of the app name. I hope my terrible ascii art can explain where the `6` comes from (the sum of the two hard-coded `3`'s in this part of the template):

```
NAME:
   <appname> - <description blah blah blah...
               we want wrapped lines to line
               up here>
^-^^-------^^-^
 |    |      |
 |    |      another hardcoded 3
 |    |
 |   length of app name
 |
hardcoded 3

```

**Comment by xordspar0** [comment]
Is this ready to be merged?

**Comment by mostynb** [comment]
> Is this ready to be merged?

It's stalled, waiting for review.

File: help.go
Comment (rliebz) [inline help.go:45]: It's not entirely clear from what's exported here, but making a change to the signature of these functions is unfortunately a breaking change, as the exported variables `HelpPrinter` and `HelpPrinterCustom` both use these signatures. [One of the CLIs I've written](https://github.com/rliebz/tusk/blob/e58ee3109f2f99972a35a054329787aa411c6357/appcli/help.go#L19-L22), for example, would no longer compile if this change were merged in.

I love the feature idea, but making a breaking change to implement it is a non-starter.

**Review by rliebz — CHANGES_REQUESTED** [review]

File: help.go
Comment (mostynb) [reply help.go]: Pushed an unpolished fixup which drops the `wrapAt` parameter, and instead allows the client to specify a "wrap" function to `HelpPrinterCustom`. If this seems reasonable I can add documentation, squash and rebase and perform some general cleanup etc.

Comment (asahasrabuddhe) [reply help.go]: This seems to be going in the correct direction to me

File: template.go
Comment (rliebz) [inline template.go:7]: Looks like we're wrapping most of the fields in the template, but not all of the fields. This PR also omits wrapping for command help and sub-command help.

Is there a reason _NOT_ to expand wrapping to all sections of all templates?

File: template.go
Comment (rliebz) [reply template.go:10]: I'm satisfied with this answer.

File: help.go
Comment (rliebz) [inline help.go:418]: Probably not a big deal, but this function name is shadowed a few times in the other functions being introduced.

**Review by rliebz — COMMENTED** [review]
Thanks for the contribution! 

This is a sweet feature, just had a few minor comments.

File: help.go
Comment (mostynb) [reply help.go]: Thanks- removed.

File: help.go
Comment (mostynb) [reply help.go]: I lifted this up into the calling function, since the same value was also used there multiple times.

File: template.go
Comment (mostynb) [reply template.go:7]: Mostly because I don't know all the variations that need to be tested. I suspect there will be some fields that aren't worth trying to wrap because they're only a problem in unusual cases, but command and sub-command help are worth supporting.

I will try to expand this a bit.

File: help.go
Comment (mostynb) [reply help.go:437]: If I understand correctly it should be safe, because there's only one `offset` function, and the other `offset`'s are `int` variables, but maybe it could be made less confusing. I could rename all the `offset` variables to `off` if you prefer?

Comment (rliebz) [reply help.go:437]: The current iteration doesn't seem to cause any problems, and doesn't seem to set off any linters I know of.

I think it should be alright as is, and renaming `offset` should be fine as well. Up to you.

**Comment by mostynb** [comment]
I was looking into the conflicting files, and noticed that #1175 has landed, which implements a subset of the feature here.

Force-pushed a squashed version of this with two fixes appended, which makes `TestWrappedHelp` also test user specifed indenting, and another change which makes `Test_ShowAppHelp_MultiLineDescription` on master pass if I revert the code changes from be9c0378066dc5baf84a56005f90317cca6ecd84 and then apply this PR.

How should we proceed?

**Comment by rliebz** [comment]
> I was looking into the conflicting files, and noticed that #1175 has landed, which implements a subset of the feature here.
> 
> Force-pushed a squashed version of this with two fixes appended, which makes `TestWrappedHelp` also test user specifed indenting, and another change which makes `Test_ShowAppHelp_MultiLineDescription` on master pass if I revert the code changes from [be9c037](https://github.com/urfave/cli/commit/be9c0378066dc5baf84a56005f90317cca6ecd84) and then apply this PR.
> 
> How should we proceed?

Those changes were contributed because someone needed them, and then tagged in a release. I think reverting them is off the table.

If you're still interested in getting this change in, you'll need to resolve conflicts. I'm totally fine if we change the template—I think folks using the default template are just signing up for whatever we decide makes sense. That said, the new template functions (indent, nindent, and trim) are now part of the public API, so those are something we shouldn't be removing.

**Comment by mostynb** [comment]
> Those changes were contributed because someone needed them, and then tagged in a release. I think reverting them is off the table.
> 
> If you're still interested in getting this change in, you'll need to resolve conflicts. I'm totally fine if we change the template—I think folks using the default template are just signing up for whatever we decide makes sense. That said, the new template functions (indent, nindent, and trim) are now part of the public API, so those are something we shouldn't be removing.

Pushed a rebased and squashed version, which undoes the template changes from be9c0378066dc5baf84a56005f90317cca6ecd84, but leaves the `indent`, `nindent` and `trim` functions in the function map for anyone who started using them with a custom template.

**Comment by mostynb** [comment]
Updated the template and added some tests to wrap commands and subcommands. There are probably other items that we can improve, but I think this is a good start.

**Comment by mostynb** [comment]
How does this look now?

**Comment by mostynb** [comment]
Ping.

**Comment by meatballhat** [comment]
@mostynb Sorry for the delay!

**Comment by meatballhat** [comment]
@mostynb Are you up for resolving the conflicts and moving this work forward? No worries if not. I'm happy to pick it up ❤️

**Comment by mostynb** [comment]
I'll take a look and let you know if I run into trouble.

**Comment by mostynb** [comment]
@meatballhat: the failing test also fails for me locally on the tip of main. The CI failure is also visible in this typo fix PR: #1402

**Comment by meatballhat** [comment]
@mostynb this looks like the problem of the docs tests running against latest release instead of current working tree :weary:

bumping the priority of that work now...

**Comment by meatballhat** [comment]
@mostynb Attempted fix for the test-docs step at #1403

**Comment by mostynb** [comment]
@meatballhat: CI is green after rebasing on main.

File: help.go
Comment (dearchap) [inline help.go:306]: I think it is better to define a type for this function signature.

Comment (mostynb) [reply help.go:306]: I'm not quite sure what you mean- are you suggesting that we export a `WrapAtFunc` type alias, and then only use `customFuncs["wrapAt"]` if it is created as a `WrapAtFunc` explicitly?

Comment (mostynb) [reply help.go:306]: Gentle ping on this :)

Comment (dearchap) [reply help.go:306]: Yes correct

Comment (mostynb) [reply help.go:306]: This requires some boilerplate for the caller, but I'm not sure how it makes things better. How does this help? I couldn't find any examples of this pattern in a quick web search, but I'm not sure what to call this so maybe I'm missing something obvious.

Comment (meatballhat) [reply help.go:306]: fwiw I'm happy to defer resolution of this to a later PR, so I'm going to move forward with merging the current work 👍🏼

**Review by meatballhat — APPROVED** [review]

**Comment by tonglil** [comment]
Is there any docs for this?

**Comment by mostynb** [comment]
> Is there any docs for this?

Perhaps it could be more discoverable, but you need to set HelpPrinterCustom or HelpPrinter to add a `wrapAt` function to the funcMap. eg:
https://github.com/urfave/cli/pull/1119/files#diff-4d3f36d62cad94424544bb84fc93b33b73ffe3491a4ddca81a7f825539afc19cR1161-R1169

**Pull request merged**
